File: graffiti/src/main/java/com/ue/graffiti/ui/PenStyleAdapter.java
package com.ue.graffiti.ui;

import android.app.Activity;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.RadioButton;
import android.widget.TextView;

import com.ue.adapterdelegate.BaseAdapterDelegate;
import com.ue.adapterdelegate.DelegationAdapter;
import com.ue.adapterdelegate.Item;
import com.ue.adapterdelegate.OnDelegateClickListener;
import com.ue.graffiti.R;
import com.ue.graffiti.base.BaseViewHolder;
import com.ue.graffiti.model.PenCatTitleItem;
import com.ue.graffiti.model.PenShapeItem;

import java.util.List;

/**
 * Created by hawk on 2018/1/17.
 */
class PenStyleAdapter extends DelegationAdapter<Item> implements OnDelegateClickListener {

    private OnDelegateClickListener mDelegateClickListener;
    private int lastShapeIndex;
    private int lastEffectIndex;

    public void setDelegateClickListener(OnDelegateClickListener delegateClickListener) {
        mDelegateClickListener = delegateClickListener;
    }

    public int getSpanCount(int position) {
        return (getItems().get(position) instanceof PenCatTitleItem) ? 4 : 1;
    }

    PenStyleAdapter(Activity activity, List<Item> items) {
        if (items != null) {
            for (int i = 0, len = items.size(); i < len; i++) {
                Item item = items.get(i);
                if (!(item instanceof PenShapeItem)) {
                    continue;
                }
                PenShapeItem shapeItem = (PenShapeItem) item;
                if (!shapeItem.isChecked) {
                    continue;
                }
                if (shapeItem.flag == PenDialog.FLAG_SHAPE) {
                    lastShapeIndex = i;
                } else {
                    lastEffectIndex = i;
                }
            }

            getItems().addAll(items);
        }

        PenItemDelegate delegate = new PenItemDelegate(activity);
        delegate.setOnDelegateClickListener(this);
        this.addDelegate(delegate);

        addDelegate(new PenTitleItemDelegate(activity));
    }

    @Override
    public void onClick(View view, int position) {
        if (position < 0 || position >= getItemCount()) {
            return;
        }

        PenShapeItem item = (PenShapeItem) getItems().get(position);
        if (item.isChecked) {
            return;
        }
        if (item.flag == PenDialog.FLAG_SHAPE) {
            ((PenShapeItem) getItems().get(lastShapeIndex)).isChecked = false;
            notifyItemChanged(lastShapeIndex, 0);
            lastShapeIndex = position;
        } else {
            ((PenShapeItem) getItems().get(lastEffectIndex)).isChecked = false;
            notifyItemChanged(lastEffectIndex, 0);
            lastEffectIndex = position;
        }
        item.isChecked = true;
        notifyItemChanged(position, 0);

        if (mDelegateClickListener != null) {
            mDelegateClickListener.onClick(view, position);
        }
    }

    private static class PenTitleItemDelegate extends BaseAdapterDelegate<Item> {
        public PenTitleItemDelegate(Activity activity) {
            super(activity, R.layout.item_pen_cat_title);
        }

        @NonNull
        @Override
        protected RecyclerView.ViewHolder onCreateViewHolder(@NonNull View itemView) {
            return new ViewHolder(itemView);
        }

        @Override
        public boolean isForViewType(@NonNull Item item) {
            return item instanceof PenCatTitleItem;
        }

        @Override
        public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, @NonNull Item item, @NonNull List payloads) {
            ((BaseViewHolder) holder).update(item);
        }

        class ViewHolder extends BaseViewHolder {
            TextView tvPenCatTitle;

            public ViewHolder(View itemView) {
                super(itemView);
                tvPenCatTitle = itemView.findViewById(R.id.tvPenCatTitle);
            }

            public void update(Object aItem) {
                PenCatTitleItem item = (PenCatTitleItem) aItem;
                tvPenCatTitle.setText(item.title);
            }
        }
    }

    private static class PenItemDelegate extends BaseAdapterDelegate<Item> {
        public PenItemDelegate(Activity activity) {
            super(activity, R.layout.item_pen_shape);
        }

        @NonNull
        @Override
        protected RecyclerView.ViewHolder onCreateViewHolder(@NonNull View itemView) {
            return new ViewHolder(itemView);
        }

        @Override
        public boolean isForViewType(@NonNull Item item) {
            return item instanceof PenShapeItem;
        }

        @Override
        public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, @NonNull Item item, @NonNull List payloads) {
            ((BaseViewHolder) holder).update(item);
        }

        class ViewHolder extends BaseViewHolder {
            RadioButton rbShapeName;

            public ViewHolder(View itemView) {
                super(itemView);
                rbShapeName = itemView.findViewById(R.id.rbShapeName);
            }

            public void update(Object aItem) {
                PenShapeItem item = (PenShapeItem) aItem;
                rbShapeName.setText(item.name);
                rbShapeName.setCompoundDrawablesWithIntrinsicBounds(0, item.image, 0, 0);
                rbShapeName.setChecked(item.isChecked);
            }
        }
    }
}

File: graffiti/src/main/java/com/ue/graffiti/ui/PenStyleAdapter.kt
package com.ue.graffiti.ui

import android.app.Activity
import android.view.View
import com.ue.adapterdelegate.AdapterDelegate
import com.ue.adapterdelegate.BaseViewHolder
import com.ue.adapterdelegate.DelegationAdapter
import com.ue.adapterdelegate.OnDelegateClickListener
import com.ue.graffiti.R
import com.ue.graffiti.model.PenCatTitleItem
import com.ue.graffiti.model.PenShapeItem
import kotlinx.android.synthetic.main.gr_item_pen_cat_title.view.*
import kotlinx.android.synthetic.main.gr_item_pen_shape.view.*

/**
 * Created by hawk on 2018/1/17.
 */
internal class PenStyleAdapter(activity: Activity, mItems: MutableList<Any>?) : DelegationAdapter<Any>(), OnDelegateClickListener {
    var delegateClickListener: OnDelegateClickListener? = null
    private var lastShapeIndex = 0
    private var lastEffectIndex = 0

    fun getSpanCount(position: Int): Int {
        return if (items[position] is PenCatTitleItem) 4 else 1
    }

    init {
        mItems?.apply {
            for (i in mItems.indices) {
                val item = mItems[i] as? PenShapeItem ?: continue
                if (item.isChecked) {
                    if (item.flag == PenDialog.FLAG_SHAPE) lastShapeIndex = i
                    else lastEffectIndex = i
                }
            }
            items.addAll(this)
        }
        addDelegate(PenTitleItemDelegate(activity))
        addDelegate(PenItemDelegate(activity).apply { delegateClickListener = this@PenStyleAdapter })
    }

    override fun onClick(view: View, position: Int) {
        val item = items[position] as PenShapeItem
        if (item.isChecked) return

        if (item.flag == PenDialog.FLAG_SHAPE) {
            (items[lastShapeIndex] as PenShapeItem).isChecked = false
            notifyItemChanged(lastShapeIndex, 0)
            lastShapeIndex = position
        } else {
            (items[lastEffectIndex] as PenShapeItem).isChecked = false
            notifyItemChanged(lastEffectIndex, 0)
            lastEffectIndex = position
        }
        item.isChecked = true
        notifyItemChanged(position, 0)

        delegateClickListener?.onClick(view, position)
    }

    private class PenTitleItemDelegate(activity: Activity) : AdapterDelegate<Any>(activity, R.layout.gr_item_pen_cat_title) {

        override fun isForViewType(item: Any) = item is PenCatTitleItem

        override fun onCreateViewHolder(itemView: View): BaseViewHolder<Any> {
            return object : BaseViewHolder<Any>(itemView) {
                private val tvPenCatTitle = itemView.tvPenCatTitle

                override fun updateContents(item: Any) {
                    item as PenCatTitleItem
                    tvPenCatTitle.text = item.title
                }
            }
        }
    }

    private class PenItemDelegate(activity: Activity) : AdapterDelegate<Any>(activity, R.layout.gr_item_pen_shape) {
        override fun isForViewType(item: Any) = item is PenShapeItem

        override fun onCreateViewHolder(itemView: View): BaseViewHolder<Any> {
            return object : BaseViewHolder<Any>(itemView) {
                private val rbShapeName = itemView.rbShapeName

                override fun updateContents(item: Any) {
                    item as PenShapeItem
                    rbShapeName.text = item.name
                    rbShapeName.setCompoundDrawablesWithIntrinsicBounds(0, item.image, 0, 0)
                    rbShapeName.isChecked = item.isChecked
                }
            }
        }
    }
}